libraries/rush-lib/src/logic/Git.ts

```diff
@@ -368,26 +425,307 @@ export class Git {
   }
 
   public hasUncommittedChanges(): boolean {
-    return this.getUncommittedChanges().length > 0;
+    const gitStatusEntries: Iterable<IGitStatusEntry> = this.getGitStatus();
+    for (const gitStatusEntry of gitStatusEntries) {
+      if (gitStatusEntry.kind !== 'ignored') {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  public hasUnstagedChanges(): boolean {
+    const gitStatusEntries: Iterable<IGitStatusEntry> = this.getGitStatus();
+    for (const gitStatusEntry of gitStatusEntries) {
+      if (
+        gitStatusEntry.kind === 'untracked' ||
+        (gitStatusEntry as IChangedGitStatusEntry).unstagedChangeType !== undefined
+      ) {
+        return true;
+      }
+    }
+
+    return false;
   }
 
   /**
    * The list of files changed but not committed
    */
   public getUncommittedChanges(): ReadonlyArray<string> {
-    const changes: string[] = [];
-    changes.push(...this._getUntrackedChanges());
-    changes.push(...this._getDiffOnHEAD());
+    const result: string[] = [];
+    const gitStatusEntries: Iterable<IGitStatusEntry> = this.getGitStatus();
+    for (const gitStatusEntry of gitStatusEntries) {
+      if (gitStatusEntry.kind !== 'ignored') {
+        result.push(gitStatusEntry.path);
+      }
+    }
 
-    return changes.filter((change) => {
-      return change.trim().length > 0;
-    });
+    return result;
   }
 
   public getTagSeparator(): string {
     return this._rushConfiguration.gitTagSeparator || DEFAULT_GIT_TAG_SEPARATOR;
   }
 
+  public *getGitStatus(): Iterable<IGitStatusEntry> {
+    const gitPath: string = this.getGitPathOrThrow();
+    // See Git.test.ts for example output
+    const output: string = this._executeGitCommandAndCaptureOutput(gitPath, [
+      'status',
+      '--porcelain=2',
+      '--null'
+    ]);
+
+    // State machine for parsing a git status entry
+    // See reference https://git-scm.com/docs/git-status?msclkid=1cff552bcdce11ecadf77a086eded66c#_porcelain_format_version_2
+
+    const enum GitStatusEntryState {
+      mode,
+      changeType,
+      headFileMode,
+      indexFileMode,
+      stage1FileMode,
+      stage2FileMode,
+      stage3FileMode,
+      worktreeFileMode,
+      headObjectName,
+      indexObjectName,
+      stage1ObjectName,
+      stage2ObjectName,
+      stage3ObjectName,
+      submoduleState,
+      renameOrCopyScore,
+      path
+    }
+
+    let pos: number = 0;
+    let state: GitStatusEntryState | undefined = GitStatusEntryState.mode;
+
+    let isRenamedOrCopied: boolean = false;
+    let isUnmerged: boolean = false;
+    let currentObject: Partial<IGitStatusEntry> = {};
```

Comment:
For large diffs the pattern of creating an empty object and adding fields to it one-by-one is going to degrade performance significantly vs. switching on the mode field and running a parser that creates an object literal directly.

Reply:
Also, 4 different functions:
- `parseUntrackedEntry`
- `parseAddModifyOrDeleteEntry`
- `parseCopyOrRenameEntry`
- `parseUnmergedEntry`
is a lot easier to read than the state machine